File: python/src/character_factory.py
from character import Character


class CharacterFactory:
    @staticmethod
    def strange_characters():
        joyce = Character("Joyce", "Byers", isMonster=False)
        jim = Character("Jim", "Hopper", isMonster=False)
        mike = Character("Mike", "Wheeler", isMonster=False)
        eleven = Character("Eleven", isMonster=False)
        dustin = Character("Dustin", "Henderson", isMonster=False)
        lucas = Character("Lucas", "Sinclair", isMonster=False)
        nancy = Character("Nancy", "Wheeler", isMonster=False)
        jonathan = Character("Jonathan", "Byers", isMonster=False)
        will = Character("Will", "Byers", isMonster=False)
        karen = Character("Karen", "Wheeler", isMonster=False)
        holly = Character("Holly", "Wheeler", isMonster=False)
        steve = Character("Steve", "Harrington", isMonster=False)
        mindflayer = Character("Mindflayer", isMonster=True)
        demagorgon = Character("Demagorgon", isMonster=True)
        demadog = Character("Demadog", isMonster=True)
        mrwhatsit = Character("MrWhatsit", isMonster=True)

        joyce._add_child(jonathan)
        joyce._add_child(will)
        jim._add_child(eleven)
        karen._add_child(nancy)
        karen._add_child(mike)
        karen._add_child(holly)

        eleven._nemesis = demagorgon
        will._nemesis = mindflayer
        dustin._nemesis = demadog
        holly._nemesis = mrwhatsit

        return [
            joyce,
            jim,
            mike,
            will,
            eleven,
            dustin,
            lucas,
            nancy,
            jonathan,
            mindflayer,
            demagorgon,
            demadog,
            karen,
            holly,
            mrwhatsit,
            steve,
        ]

    @staticmethod
    def wheeler_family():
        mike = Character("Mike", "Wheeler", isMonster=False)
        nancy = Character("Nancy", "Wheeler", isMonster=False)
        karen = Character("Karen", "Wheeler", isMonster=False)
        holly = Character("Holly", "Wheeler", isMonster=False)

        karen._add_child(nancy)
        karen._add_child(mike)
        karen._add_child(holly)

        return [
            nancy,
            karen,
            mike,
            holly,
        ]

File: cpp/src/CharacterFactory.cpp
#include "CharacterFactory.hpp"

namespace Characters {

std::vector<std::shared_ptr<Character>> CharacterFactory::StrangeCharacters() {
    auto joyce = std::make_shared<Character>("Joyce", "Byers");
    auto jim = std::make_shared<Character>("Jim", "Hopper");
    auto mike = std::make_shared<Character>("Mike", "Wheeler");
    auto eleven = std::make_shared<Character>("Eleven");
    auto dustin = std::make_shared<Character>("Dustin", "Henderson");
    auto lucas = std::make_shared<Character>("Lucas", "Sinclair");
    auto nancy = std::make_shared<Character>("Nancy", "Wheeler");
    auto jonathan = std::make_shared<Character>("Jonathan", "Byers");
    auto will = std::make_shared<Character>("Will", "Byers");
    auto karen = std::make_shared<Character>("Karen", "Wheeler");
    auto holly = std::make_shared<Character>("Holly", "Wheeler");
    auto steve = std::make_shared<Character>("Steve", "Harrington");
    auto mindflayer = std::make_shared<Character>("Mindflayer", std::nullopt, true);
    auto demagorgon = std::make_shared<Character>("Demagorgon", std::nullopt, true);
    auto demadog = std::make_shared<Character>("Demadog", std::nullopt, true);
    auto mrwhatsit = std::make_shared<Character>("MrWhatsit", std::nullopt, true);

    joyce->AddChild(jonathan.get());
    joyce->AddChild(will.get());
    jim->AddChild(eleven.get());
    karen->AddChild(nancy.get());
    karen->AddChild(mike.get());
    karen->AddChild(holly.get());

    eleven->SetNemesis(demagorgon.get());
    will->SetNemesis(mindflayer.get());
    dustin->SetNemesis(demadog.get());
    holly->SetNemesis(mrwhatsit.get());

    return {
            joyce,
            jim,
            mike,
            will,
            eleven,
            dustin,
            lucas,
            nancy,
            jonathan,
            mindflayer,
            demagorgon,
            demadog,
            karen,
            holly,
            steve,
            mrwhatsit
    };
}

std::vector<std::shared_ptr<Character>> CharacterFactory::WheelerFamily() {
    auto karen = std::make_shared<Character>("Karen", "Wheeler");
    auto mike = std::make_shared<Character>("Mike", "Wheeler");
    auto nancy = std::make_shared<Character>("Nancy", "Wheeler");
    auto holly = std::make_shared<Character>("Holly", "Wheeler");

    karen->AddChild(nancy.get());
    karen->AddChild(mike.get());
    karen->AddChild(holly.get());

    return {
            karen,
            mike,
            nancy,
            holly
    };
}

}
